collector/src/main.rs

```diff
@@ -1141,7 +1166,100 @@ fn main_result() -> anyhow::Result<i32> {
 
             Ok(0)
         }
+        Commands::Download(cmd) => {
+            let target_dir = get_downloaded_crate_target(&benchmark_dir, &cmd);
+            check_target_dir(&target_dir, cmd.overwrite)?;
+
+            match cmd.command {
+                DownloadSubcommand::Git { url } => download_from_git(&target_dir, &url)?,
+                DownloadSubcommand::Crate { krate, version } => {
+                    download_from_crates_io(&target_dir, &krate, &version)?
+                }
+            };
+            println!("Benchmark stored at {}", target_dir.display());
+            Ok(0)
+        }
+    }
+}
+
+fn check_target_dir(target_dir: &Path, overwrite: bool) -> anyhow::Result<()> {
+    if target_dir.exists() {
+        if overwrite {
+            std::fs::remove_dir_all(&target_dir).expect(&format!(
+                "Cannot remove previous directory at {}",
+                target_dir.display()
+            ));
+        } else {
+            return Err(anyhow::anyhow!(
+                "Directory {} already exists",
+                target_dir.display()
+            ));
+        }
     }
+    Ok(())
+}
+
+fn get_downloaded_crate_target(benchmark_dir: &Path, cmd: &DownloadCommand) -> PathBuf {
+    let name = cmd.name.clone().unwrap_or_else(|| match &cmd.command {
+        DownloadSubcommand::Git { url } => url
+            .trim_end_matches("/")
+            .trim_end_matches(".git")
+            .split("/")
+            .last()
+            .expect("Crate name could not be determined from git URL")
+            .to_string(),
+        DownloadSubcommand::Crate { krate, version } => format!("{krate}-{version}"),
+    });
+    PathBuf::from(benchmark_dir).join(name)
+}
+
+fn download_from_git(target: &Path, url: &str) -> anyhow::Result<()> {
+    let tmpdir = tempfile::TempDir::new().unwrap();
+    Command::new("git")
+        .arg("clone")
+        .arg(url)
+        .arg(tmpdir.path())
+        .status()
+        .expect("Git clone failed");
+    generate_lockfile(tmpdir.path());
+    execute::rename(&tmpdir, &target)?;
+    Ok(())
+}
+
+fn download_from_crates_io(target_dir: &Path, krate: &str, version: &str) -> anyhow::Result<()> {
+    let url = format!("https://crates.io/api/v1/crates/{krate}/{version}/download");
+    let response = reqwest::blocking::get(url)
+        .expect("Cannot download crate")
+        .error_for_status()?;
+
+    let data = flate2::read::GzDecoder::new(response);
+    let mut archive = tar::Archive::new(data);
+
+    let tmpdir = tempfile::TempDir::new().unwrap();
+    archive.unpack(&tmpdir)?;
+    let unpacked_dir = std::fs::read_dir(&tmpdir.path())
+        .unwrap()
+        .next()
+        .unwrap()
+        .unwrap()
+        .path();
```

Comment:
What is this read_dir trying to do? It looks like it's getting the path of the first entry in the directory, but I don't understand what the goal is...

Reply:
When we download the crate and unzip it into e.g. `/tmp/foo` it will create a directory `/tmp/foo/syn-1.0.0` (the crate is not at the root of the [g]zipped file, but inside a directory). So I want to find the expanded directory and copy its contents to the target directory.

Reply:
Good place for a brief comment :)

Reply:
I might also suggest guessing the format of the crate name (name-version) and just moving that. Alternatively, we might consider unpacking directly into collector/benchmarks, since that seems like it'll largely just work.

Reply:
Ok, sounds good to me.